### PR TITLE
151 copy mode for the aegtree

### PR DESCRIPTION
**Pull request by AnushaTiwari5**
Copy works, includes reviewed changes from #172

### CODE REVIEW

File: src/AEG/CutNode.ts [hunk 1]
```diff
@@ -122,6 +122,71 @@ export class CutNode {
         }
     }
 
+    /**
+     * Recursive method to return the lowest level node containing the given point.
+     * @param incomingPoint The given point on the canvas.
+     * @returns The lowest node containing the node on the tree.
```
Comment (James-Oswald) [inline src/AEG/CutNode.ts:128]: You need to document under what circumstances this function returns null, as you claim it does in the return value.

File: src/AEG/CutNode.ts [hunk 2]
```diff
@@ -122,6 +122,71 @@ export class CutNode {
...
+        }
+
+        //None of the children contain the point, so this is the lowest node containing the point.
+        return this;
+    }
+
+    /**
+     * Recursive method to return the parent of the lowest node containing the given point.
+     * @param incomingPoint The given point on the canvas.
+     * @returns The parent of the lowest level node.
```
Comment (James-Oswald) [inline src/AEG/CutNode.ts:157]: Likewise, document when null is returned.

File: src/CopySingleMode.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            const tempAtom: AtomNode = alterAtom(currentNode, moveDifference);
+
+            ctx.clearRect(0, 0, canvas.width, canvas.height);
+            redrawCut(tree.sheet, offset);
+
+            if (tree.canInsert(tempAtom)) {
+                drawAtom(tempAtom, legalColor(), true);
+            } else {
+                drawAtom(tempAtom, illegalColor(), true);
+            }
```
Comment (James-Oswald) [inline src/CopySingleMode.ts:90]: Not directly relevant to this PR, but during code cleanup. it should be made clear what this bool to draw atom is and give it a better parameter name than `currentAtom` which in my mind is an atom.

File: src/EditModeUtils.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            }
+        }
+    } else if (incomingNode instanceof AtomNode) {
+        const tempAtom: AtomNode = alterAtom(incomingNode, change);
+
+        tree.insert(tempAtom);
+    }
+}
+
+export function alterCut(originalCut: CutNode, difference: Point): CutNode {
```
Comment (James-Oswald) [inline src/EditModeUtils.ts:99]: documentation?

File: src/EditModeUtils.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                originalCut.ellipse.radiusX,
+                originalCut.ellipse.radiusY
+            )
+        );
+    } else {
+        throw new Error("Cannot alter the position of a cut without an ellipse.");
+    }
+}
+
+export function alterAtom(originalAtom: AtomNode, difference: Point) {
```
Comment (James-Oswald) [inline src/EditModeUtils.ts:116]: documentation?

File: src/MoveSingleMode.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+ * Otherwise enters the renters currentNode into the tree at the original point.
+ * @param event The mouse up event while in moveSingle mode
+ */
+export function moveSingleMouseUp(event: MouseEvent) {
+    if (legalNode) {
+        const moveDifference: Point = new Point(
+            event.x - startingPoint.x,
+            event.y - startingPoint.y
+        );
+        //
```
Comment (James-Oswald) [inline src/MoveSingleMode.ts:114]: Stray `//`

File: src/AEG/CutNode.ts (hunk 1 above)
Comment (DawnTheWitch) [reply src/AEG/CutNode.ts:128]: This is all my bad, I kind of just focused on adding the reviews without really thinking, will add all of these when I wake up tomorrow.

**Review by DawnTheWitch — APPROVED** [review]

**Pull request merged**